upp/stages/resampling.py

```diff
@@ -29,6 +30,63 @@ def safe_divide(a, b):
     return np.divide(a, b, out=np.zeros_like(a), where=b != 0)
 
 
+def subdivide_bins(bins, n=2):
+    comb_list = [np.array((bins[0],))]
+    for i in range(len(bins) - 1):
+        comb_list.append(np.linspace(bins[i], bins[i + 1], n + 1)[1:])
+    return np.concatenate(comb_list)
+
+
+def upscale_array(array, upscl, order=3, mode="nearest", nornalise=False, positive=True):
+    # upscl must be integer
+    xs = []
+    for d in array.shape:
+        n_bins = d
+        points = np.linspace(-0.5 + 1 / 2 / upscl, n_bins - 0.5 - 1 / 2 / upscl, n_bins * upscl)
+        xs.append(points)
+
+    # return the smoothed array
+    xy = np.meshgrid(*xs, indexing="ij")
+    smoothed = ndimage.map_coordinates(array, xy, order=order, mode=mode)
+    if positive:
+        smoothed[smoothed < 0] = 0
+    if nornalise:
+        smoothed = smoothed / smoothed.sum()
+    return smoothed
+
+
+def upscale_array_regionally(
```

Comment:
Should we put this function and the above functions into the hist class as they were before? They could probably also do with some type hints and at least brief docstrings

Reply:
I don't think we have to put them in the hist class as I apply them generally to any array. As you see in PDF method I apply them not to the hits but to the ratio of the histograms. So I would rather put them in something like "utils". What do you think @samvanstroud ?

Reply:
Thanks a lot @IvanOleksiyuk, that sounds good to me, better to call it `interpolation.py` or `smoothing.py` or something a bit more related to the functionality if possible.

Reply:
I moved the functions to interpolation.py